In [9]:
import pandas as pd

Вывод первых и последних строк

→ Часто бывает такое, что вывести на экран все строки таблицы является ресурсозатратной операцией, а иногда и вовсе не представляется возможным. В большинстве случаев для того, чтобы понять структуру DataFrame и удостовериться, что таблица подгрузилась верно, достаточно вывести несколько первых или последних строк.

Для этого у DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5).

In [8]:
#Выведем первые пять строк нашей таблицы:
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
display(melb_data.head())

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [10]:
#Следующий код выведет семь последних строк нашей таблицы:
melb_data.tail(7)

index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577          Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578        Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579        Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
13573     1980.0          NaN  -37.89327   144.64789   
13574     1970.0          NaN  -37.67917   144.89390   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13573        Western Metropolitan        16166.0  -37.89327, 144.64789  
13574       Northern Metropolitan         2474.0   -37.67917, 144.8939  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[7 rows x 23 columns]

Данный метод выводит:

информацию об индексах;
информацию об общем количестве столбцов;
таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

In [11]:
#Размерность таблицы
melb_data.shape

(13580, 23)

Из вывода метода info() становится понятно, что в нашей таблице есть столбец CouncilArea с пропущенными значениями — количество непустых значений в столбце меньше, чем количество строк в таблице (12211 < 13580).

Пустыми, или пропущенными, значениями называются значения в ячейках таблицы, которые не заполнены по какой-либо причине, то есть на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number).

→ Мы будем говорить о работе с пропусками в отдельном модуле, посвящённом очистке данных, а пока что не будем обращать на них внимания.

Если вам всё же любопытно, как можно работать с такими значениями в Pandas, то вы можете посмотреть информацию здесь.

https://pythonru.com/biblioteki/not-a-number-vse-o-nan-pd-5

In [23]:
#Получение информации о столбцах
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

Например, данные в столбцах, которые отражают количество, должны, по идее, выражаться целым числом (Car, Bedroom, Bathroom и Propertyсount), однако кодируются float64 — числом с плавающей запятой размером 64 бита.

Наконец, данные в столбце с годом постройки (YearBuilt) также представлены в формате чисел с плавающей точкой.

Чтобы исправить это, можно воспользоваться методом astype(), который позволяет преобразовать тип данных столбца:

In [7]:
#В данном коде мы при помощи метода astype() последовательно переопределяем столбцы на них же самих, 
# только с изменённым типом данных: int64 — целочисленное число размером 64 бита.
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
melb_data.info()

NameError: name 'melb_data' is not defined

Чтобы не увязнуть в обилии информации, выведем на экран значение статистических параметров только для столбцов Distance (расстояние от объекта недвижимости до центра Мельбурна), BuildingArea (площадь здания) и Price (цена объекта):

In [49]:
melb_data.describe().loc[:, ['Distance', 'BuildingArea' , 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

→ На самом деле метод describe() можно применять не только к числовым признакам. С помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию.

Например, для типа данных object метод describe() возвращает DataFrame, в котором указаны:

количество непустых строк (count);
количество уникальных значений (unique);
самое частое значение — мода —  (top);
частота — объём использования — этого значения (freq) для каждого столбца типа object исходной таблицы.

In [10]:
melb_data.describe(include=['object'])

Suburb       Address   Type Method SellerG        Date CouncilArea  \
count       13580         13580  13580  13580   13580       13580       12211   
unique        314         13378      3      5     268          58          33   
top     Reservoir  5 Charles St      h      S  Nelson  27/05/2017    Moreland   
freq          359             3   9449   9022    1565         473        1163   

                   Regionname         Coordinates  
count                   13580               13580  
unique                      8               13097  
top     Southern Metropolitan  -37.8361, 144.9966  
freq                     4695                  12

Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота.

Рассмотрим работу value_counts() на примере столбца с названиями районов:

In [11]:
melb_data['Regionname'].value_counts()

Regionname
Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: count, dtype: int64

Чтобы сделать вывод более интерпретируемым и понятным, можно воспользоваться параметром normalize. При установке значения этого параметра на True результат будет представляться в виде доли (относительной частоты):

In [12]:
melb_data['Regionname'].value_counts(normalize=True)

Regionname
Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: proportion, dtype: float64

In [18]:
#Задача 6.1
#Дана таблица data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B']). Что выведет код print(data.shape)?
data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B'])
print(data.shape)

(3, 2)


In [20]:
#Задача 6.1
#Какой из нижеперечисленных вариантов кода позволяет получить последние 15 строк DataFrame melb_data?
melb_data.tail(15)

index          Suburb             Address  Rooms Type      Price  \
13565  13565         Vermont      73 Nurlendi Rd      4    h  1263500.0   
13566  13566   Vermont South       9 Winswood Cl      4    h  1250000.0   
13567  13567        Viewbank       149 Graham Rd      5    h  1316000.0   
13568  13568        Wantirna  16 chesterfield Ct      4    h   951000.0   
13569  13569  Wantirna South        12 Armagh Cr      4    h  1323000.0   
13570  13570  Wantirna South       34 Fewster Dr      3    h   970000.0   
13571  13571  Wantirna South          15 Mara Cl      4    h  1330000.0   
13572  13572        Watsonia       76 Kenmare St      2    h   650000.0   
13573  13573        Werribee         5 Nuragi Ct      4    h   635000.0   
13574  13574     Westmeadows          9 Black St      3    h   582000.0   
13575  13575   Wheelers Hill        12 Strada Cr      4    h  1245000.0   
13576  13576    Williamstown       77 Merrett Dr      3    h  1031000.0   
13577  13577    Williamstown         83 Power St      3    h  1170000.0   
13578  13578    Williamstown        96 Verdon St      4    h  2500000.0   
13579  13579      Yarraville          6 Agnes St      4    h  1285000.0   

      Method        SellerG        Date  Distance  ...  Car  Landsize  \
13565      S          Barry  26/08/2017      17.2  ...    2     791.0   
13566     PI         Philip  26/08/2017      17.2  ...    2     986.0   
13567     SP         Nelson  26/08/2017       8.9  ...    3     696.0   
13568      S            Ray  26/08/2017      14.7  ...    2     704.0   
13569      S          Barry  26/08/2017      14.7  ...    2     551.0   
13570      S          Barry  26/08/2017      14.7  ...    2     674.0   
13571      S          Barry  26/08/2017      14.7  ...    2     717.0   
13572     PI       Morrison  26/08/2017      14.5  ...    1     210.0   
13573      S  hockingstuart  26/08/2017      14.7  ...    1     662.0   
13574      S            Red  26/08/2017      16.5  ...    2     256.0   
13575      S          Barry  26/08/2017      16.7  ...    2     652.0   
13576     SP       Williams  26/08/2017       6.8  ...    2     333.0   
13577      S          Raine  26/08/2017       6.8  ...    4     436.0   
13578     PI        Sweeney  26/08/2017       6.8  ...    5     866.0   
13579     SP        Village  26/08/2017       6.3  ...    1     362.0   

       BuildingArea  YearBuilt  CouncilArea  Lattitude  Longtitude  \
13565         126.0       1970          NaN  -37.84467   145.19276   
13566         126.0       1970          NaN  -37.84679   145.20051   
13567         126.0       1970          NaN  -37.73501   145.08341   
13568         200.0       1981          NaN  -37.85639   145.23536   
13569         126.0       1970          NaN  -37.87608   145.22390   
13570         126.0       1970          NaN  -37.88360   145.22805   
13571         191.0       1980          NaN  -37.86887   145.22116   
13572          79.0       2006          NaN  -37.70657   145.07878   
13573         172.0       1980          NaN  -37.89327   144.64789   
13574         126.0       1970          NaN  -37.67917   144.89390   
13575         126.0       1981          NaN  -37.90562   145.16761   
13576         133.0       1995          NaN  -37.85927   144.87904   
13577         126.0       1997          NaN  -37.85274   144.88738   
13578         157.0       1920          NaN  -37.85908   144.89299   
13579         112.0       1920          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13565        Eastern Metropolitan           4181  -37.84467, 145.19276  
13566        Eastern Metropolitan           4280  -37.84679, 145.20051  
13567        Eastern Metropolitan           2698  -37.73501, 145.08341  
13568        Eastern Metropolitan           5424  -37.85639, 145.23536  
13569        Eastern Metropolitan           7082   -37.87608, 145.2239  
13570        Eastern Metropolitan           7082   -37.8836, 145.22805  
13571

In [78]:
#Какой тип данных имеет столбец Coordinates?
melb_data['Coordinates']

0          -37.7996, 144.9984
1          -37.8079, 144.9934
2          -37.8093, 144.9944
3          -37.7969, 144.9969
4          -37.8072, 144.9941
                 ...         
13575    -37.90562, 145.16761
13576    -37.85927, 144.87904
13577    -37.85274, 144.88738
13578    -37.85908, 144.89299
13579    -37.81188, 144.88449
Name: Coordinates, Length: 13580, dtype: object

In [77]:
#количество непустых строк (count);
#количество уникальных значений (unique);
#самое частое значение — мода —  (top);
#частота — объём использования — этого значения (freq) для каждого столбца типа object исходной таблицы.

melb_data.describe(include=['object'])

Suburb       Address   Type Method SellerG        Date CouncilArea  \
count       13580         13580  13580  13580   13580       13580       12211   
unique        314         13378      3      5     268          58          33   
top     Reservoir  5 Charles St      h      S  Nelson  27/05/2017    Moreland   
freq          359             3   9449   9022    1565         473        1163   

                   Regionname         Coordinates  
count                   13580               13580  
unique                      8               13097  
top     Southern Metropolitan  -37.8361, 144.9966  
freq                     4695                  12

In [14]:
#Сколько столбцов после преобразования типов имеет целочисленный тип данных?
melb_data.dtypes

index              int64
Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode           int64
Bedroom          float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
Coordinates       object
dtype: object

In [13]:
#Выберите верные утверждения в отношении таблицы melb_data
melb_data.describe().loc[:, ['Distance', 'BuildingArea', 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

In [15]:
#Выберите верные утверждения в отношении таблицы melb_data:
display(melb_data['Regionname'].value_counts())
display(melb_data['Coordinates'].value_counts())
display(melb_data['Suburb'].value_counts())
x = melb_data['Type'].value_counts()
print(x)


Regionname
Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: count, dtype: int64

Coordinates
-37.8361, 144.9966      12
-37.8035, 144.985        8
-37.8424, 144.9449       6
-37.8127, 145.0094       5
-37.8415, 144.9412       5
                        ..
-37.76684, 145.11309     1
-37.7663, 145.1132       1
-37.76078, 145.10696     1
-37.7587, 144.99966      1
-37.7969, 144.9969       1
Name: count, Length: 13097, dtype: int64

Suburb
Reservoir         359
Richmond          260
Bentleigh East    249
Preston           239
Brunswick         222
                 ... 
Attwood             1
Wallan              1
New Gisborne        1
Plumpton            1
Monbulk             1
Name: count, Length: 314, dtype: int64

Type
h    9449
u    3017
t    1114
Name: count, dtype: int64

In [20]:
#Сколько процентов от общего количества домов составляют таунхаусы (тип объекта — t)?
#Ответ округлите до целого числа. Не указывайте знак %.
x = melb_data['Type'].value_counts()
procent = x['t']/(sum(x)/100)
print(procent)

8.203240058910161
